internal/backend/bn254/groth16/solidity.go

```diff
@@ -217,10 +263,42 @@ contract Verifier {
         // Make sure that every input is less than the snark scalar field
         for (uint256 i = 0; i < input.length; i++) {
             require(input[i] < SNARK_SCALAR_FIELD,"verifier-gte-snark-scalar-field");
-            vk_x = Pairing.plus(vk_x, Pairing.scalar_mul(vk.IC[i + 1], input[i]));
         }
 
-        vk_x = Pairing.plus(vk_x, vk.IC[0]);
+        VerifyingKey memory vk = verifyingKey();
+
+        // Compute the linear combination vk_x
+        Pairing.G1Point memory vk_x = Pairing.G1Point(0, 0);
+
+        // Buffer reused for addition. Avoids memory allocations
+        uint256[4] memory add_input;
+        Pairing.G1Point memory add_return = Pairing.G1Point(0, 0);
+
+        // Buffer reused for multiplication
+        uint256[3] memory mul_input;
+        Pairing.G1Point memory mul_return = Pairing.G1Point(0, 0);
+
+
+        {{- range $i, $ki := .G1.K }}
+            {{- if gt $i 0 -}}
+                {{- $pos := sub $i 1 }}
+        mul_input[0] = uint256({{$ki.X.String}});
+        mul_input[1] = uint256({{$ki.Y.String}});
+        mul_input[2] = input[{{$pos}}];
+        iteration(mul_input, mul_return, add_input, add_return);
+            {{- end -}}
+        {{- end }}
+
+        {{- range $i, $ki := .G1.K -}}
+            {{- if lt $i 1}}
+        // last
+        add_input[0] = add_return.X;
+        add_input[1] = add_return.Y;
+        add_input[2] = uint256({{$ki.X.String}});
+        add_input[3] = uint256({{$ki.Y.String}});
+        Pairing.plus_raw(add_input, add_return);
+             {{- end -}}
+        {{- end }}
```

Comment:
test were failing because; vk_x is used in the pairing computation, while the result from add_return is not. 
Will merge and do the quick fix on develop